types/src/auction/detail.rs

```diff
@@ -66,3 +70,69 @@ pub(crate) fn process_unbond_requests<P: Auction + ?Sized>(provider: &mut P) ->
     provider.write(unbonding_purses_uref, new_unbonding_purses)?;
     Ok(())
 }
+
+/// Update delegators entry. Initialize if it doesn't exist.
+pub fn update_delegators<P>(
+    provider: &mut P,
+    validator_public_key: PublicKey,
+    delegator_public_key: PublicKey,
+    delegation_amount: U512,
+) -> Result<U512>
+where
+    P: RuntimeProvider + StorageProvider + ?Sized,
+{
+    let mut delegators = internal::get_delegators(provider)?;
+    let new_quantity = *delegators
+        .entry(validator_public_key)
+        .or_default()
+        .entry(delegator_public_key)
+        .and_modify(|delegation| *delegation += delegation_amount)
+        .or_insert_with(|| delegation_amount);
+    internal::set_delegators(provider, delegators)?;
+    Ok(new_quantity)
+}
+
+/// Update validator reward map.
+pub fn update_delegator_rewards<P>(
+    provider: &mut P,
+    validator_public_key: PublicKey,
+    rewards: impl Iterator<Item = (PublicKey, Ratio<U512>)>,
+) -> Result<Ratio<U512>>
+where
+    P: MintProvider + RuntimeProvider + StorageProvider + SystemProvider + ?Sized,
+{
+    let mut remainder = Ratio::zero();
+    let mut outer = internal::get_delegator_reward_map(provider)?;
+    let mut inner = outer.remove(&validator_public_key).unwrap_or_default();
+
+    for (delegator_key, delegator_reward) in rewards {
+        let delegator_reward_trunc = delegator_reward.to_integer();
+        inner
+            .entry(delegator_key)
+            .and_modify(|sum| *sum += delegator_reward_trunc)
+            .or_insert_with(|| delegator_reward_trunc);
+        remainder += delegator_reward.fract();
```

Comment:
It panics (see below).

I don't know what that second sentence means.

Reply:
The crate seems to cancel common factors from the fractions, but otherwise let numerator and denominator grow until they overflow. So even "small" values can, after a few arithmetic operations, end up overflowing, because both numerator and denominator get very large.

And addition, subtraction, multiplication and division can all cause them to increase.

Reply:
Is this a problem that will affect my implementation here?

Reply:
This method alone would certainly overflow if called with values where not many common factors can be canceled during the calculation.
But maybe it's actually okay, because we're only calling it with `rewards` that all have the same denominator (or rather, a small least common multiple), because they all result from dividing by the same number?

So I'm not sure; we somehow need to track of how many bits the denominator can have at any point in the code, and round before that can exceed 512. If it can't, it would be good to add comments that explain why, I think.

Reply:
Thanks, I understand your concerns better now.